host/s3.go

```diff
@@ -0,0 +1,412 @@
+	SecretAccessKey string
+	Region          string
+	Endpoint        string
+	Bucket          string
+	Prefix          string
+	FlushInterval   time.Duration
+}
+
+// IsValid checks whether the current configuration is valid or not
+// returns true if the S3 Region, the Bucket, the AccessKeyID and the Secret are set
+// returns false otherwise
+func (s3cfg *S3DSConfig) IsValid() bool {
+	return len(s3cfg.Bucket) > 0 &&
+		len(s3cfg.AccessKeyID) > 0 &&
+		len(s3cfg.SecretAccessKey) > 0 &&
+		len(s3cfg.Region) > 0 &&
+		s3cfg.FlushInterval.Nanoseconds() > 0
+}
+
+// ToAWSconfig makes a quick translation from the given user args
+// into the aws.Config struct -> ready to create the S3 client
+func (s3cfg S3DSConfig) ToAWSconfig() (*aws.Config, error) {
+	if !s3cfg.IsValid() {
+		return nil, fmt.Errorf("non valid s3 configuration, %+v", s3cfg)
+	}
+	cfg, err := config.LoadDefaultConfig(
+		context.TODO(),
+		config.WithRegion(s3cfg.Region),
+		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
+			s3cfg.AccessKeyID,
+			s3cfg.SecretAccessKey,
+			"", // empty session for now
+		)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+// TraceBatcher is an internal threathsafe buffer for EventTraces
+type traceBatcher struct {
+	sync.RWMutex
+	limit    int64 // bytes
+	traces   []*TraceEvent
+	totBytes int64
+}
+
+// NewTraceBatcher creates a new empty TraceBatcher
+func newTraceBatcher(limit int64) (*traceBatcher, error) {
+	if limit <= 0 {
+		return nil, fmt.Errorf("invalid size for the trace limitter %d", limit)
+	}
+	return &traceBatcher{
+		limit:  limit,
+		traces: make([]*TraceEvent, 0), // limitted by size, not by traces
+	}, nil
+}
+
+// addNewTrace locks the array and adds a new event to the queue
+// it also aggregates the bytes from the json format to know when the
+// buffer needs to be  flushed
+func (b *traceBatcher) addNewTrace(event *TraceEvent) {
+	// this might be rustic, but we calculate the size of the trace
+	// based on the it's JsonBytes (although we used compressed parquets)
+	jsonBytes := event.Data()
+	b.Lock()
+	b.traces = append(b.traces, event)
+	b.totBytes = b.totBytes + int64(len(jsonBytes))
+	b.Unlock()
+}
+
+func (b *traceBatcher) len() int {
+	b.RLock()
+	defer b.RUnlock()
+	return len(b.traces)
+}
+
+// isFull checks whether we've already reached the limit of
+// bytes to flush the batch of traces
+// NOTE: takes into account the Json Encoded bytes of the trace
+func (b *traceBatcher) isFull() bool {
+	b.RLock()
+	defer b.RUnlock()
+	return b.totBytes >= b.limit
+}
+
+// reset makes a copy of the existing traces
+// converting them into a parquet formatted events
+// and will return the copy ready to be submitted
+func (b *traceBatcher) reset() []ParquetTraceEvent {
+	b.Lock()
+	prevTraces := make([]ParquetTraceEvent, len(b.traces))
+	for i, trace := range b.traces {
+		prevTraces[i] = *trace.toParquet()
+	}
+	b.traces = make([]*TraceEvent, 0)
+	b.Unlock()
+	return prevTraces
```

Comment:
To avoid the data copy I _think_ (not sure) that you can do something like this:

```go
func (b *traceBatcher) reset() []ParquetTraceEvent {
	b.Lock()
	prevTraces := b.traces
	b.traces = nil
	b.Unlock()
	return prevTraces
```

I'm really not sure about this.